lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/DriverStatus.java

```diff
@@ -22,9 +22,57 @@
  * The status of the Driver.
  */
 public enum DriverStatus {
+
   PRE_INIT,
   INIT,
   RUNNING,
   SHUTTING_DOWN,
-  FAILING
+  FAILING;
+
+  /**
+   * Check if the driver is in process of shutting down (either gracefully or due to an error).
+   * @return true if the driver is shutting down (gracefully or otherwise).
+   */
+  public boolean isClosing() {
+    return this == SHUTTING_DOWN || this == FAILING;
+  }
+
+  /**
+   * Check whether a driver state transition from current state to a given one is legal.
+   * @param toStatus Destination state.
+   * @return true if transition is valid, false otherwise.
+   */
+  public boolean isLegalTransition(final DriverStatus toStatus) {
+
+    switch (this) {
+
+    case PRE_INIT:
+      switch (toStatus) {
+      case INIT:
+        return true;
+      default:
+        return false;
+      }
+
+    case INIT:
+      switch (toStatus) {
+      case RUNNING:
+        return true;
+      default:
+        return false;
+      }
+
+    case RUNNING:
+      switch (toStatus) {
+      case SHUTTING_DOWN:
+      case FAILING:
+        return true;
+      default:
+        return false;
+      }
+
+    default:
+      return false;
```

Comment:
This is a logic change, even though minor - previously this switch checked all states we consider and threw exception if current state is unconsidered. This is potentially useful - if we add another state we'll be forced to revisit this logic. I'd keep it